File: 543-diameter-of-binary-tree/diameter-of-binary-tree.ts
/**
 * Definition for a binary tree node.
 * class TreeNode {
 *     val: number
 *     left: TreeNode | null
 *     right: TreeNode | null
 *     constructor(val?: number, left?: TreeNode | null, right?: TreeNode | null) {
 *         this.val = (val===undefined ? 0 : val)
 *         this.left = (left===undefined ? null : left)
 *         this.right = (right===undefined ? null : right)
 *     }
 * }
 */

function diameterOfBinaryTree(root: TreeNode | null): number {
    let maxDiameter: number = 0

    function dfs(node: TreeNode | null): number {
        if (node === null) return 0

        const leftDepth: number = dfs(node.left)
        const rightDepth: number = dfs(node.right)
        
        maxDiameter = Math.max(maxDiameter, rightDepth + leftDepth)

        return Math.max(rightDepth, leftDepth) + 1
    }

    dfs(root)

    return maxDiameter
};

File: 543-diameter-of-binary-tree/diameter-of-binary-tree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def diameterOfBinaryTree(self, root: Optional[TreeNode]) -> int:
        max_diameter = 0

        def get_height(node):
            nonlocal max_diameter
            if not node:
                return 0

            left_height = get_height(node.left)
            right_height = get_height(node.right)

            max_diameter = max(max_diameter, left_height + right_height)

            return 1 + max(left_height, right_height)

        get_height(root)
        return max_diameter